File: src/main/java/raylras/zen/code/type/IntersectionType.java
package raylras.zen.code.type;

import raylras.zen.code.CompilationEnvironment;
import raylras.zen.code.SymbolProvider;
import raylras.zen.code.symbol.Symbol;

import java.util.Collection;
import java.util.List;
import java.util.stream.Collectors;

public class IntersectionType extends Type implements SymbolProvider {

    private final List<Type> typeList;

    public IntersectionType(List<Type> typeList) {
        this.typeList = typeList;
    }

    @Override
    public Collection<Symbol> getSymbols() {
        MemberValidator validator = new MemberValidator();
        for (Type type : typeList) {
            if (type instanceof SymbolProvider symbolProvider) {
                validator.addAll(symbolProvider);
            }
        }
        return validator.getMembers();
    }

    @Override
    public boolean isInheritedFrom(Type type) {
        return typeList.stream().anyMatch(it -> it.isInheritedFrom(type));
    }

    @Override
    public boolean isCastableTo(Type type, CompilationEnvironment env) {
        return typeList.stream().anyMatch(it -> it.isCastableTo(type, env));
    }

    public List<Type> getTypeList() {
        return typeList;
    }

    @Override
    public String toString() {
        return typeList.stream().map(Object::toString).collect(Collectors.joining(" & "));
    }

}

File: core/src/main/kotlin/raylras/intellizen/type/IntersectionType.kt
package raylras.intellizen.type

import raylras.intellizen.CompilationEnvironment
import raylras.intellizen.symbol.Symbol
import raylras.intellizen.symbol.SymbolProvider

data class IntersectionType(val typeList: List<Type>) : Type, SymbolProvider {
    override val typeName: String by lazy {
        typeList.joinToString(" & ") { it.typeName }
    }

    override val simpleTypeName: String by lazy {
        typeList.joinToString(" & ") { it.simpleTypeName }
    }

    override fun isSupertypeTo(type: Type): Boolean {
        return typeList.any { it.isSupertypeTo(type) }
    }

    override fun getSymbols(env: CompilationEnvironment?): Sequence<Symbol> {
        val validator = MemberValidator()
        for (type in typeList) {
            if (type is SymbolProvider) {
                type.getSymbols().forEach { validator.add(it) }
            }
        }
        return validator.getMembers().asSequence()
    }

    override fun toString() = typeName
}
